File: craigslist_jobs/python/craigslist_jobs.py
#!/usr/bin/env python

"""
Code to read Craigs List jobs for a given city, returning a RSS feed of jobs
matching the given keywords. 
"""

from datetime import datetime
from lxml import html
import re
import rsslib


# modify these as needed
JOBS_URL = 'http://phoenix.craigslist.org/sof/index.html'
HREF_STR = 'phoenix.craigslist.org/evl/sof/'
KEYWORDS = frozenset(['perl', 'ruby', 'python', 'postgres', 'postgresql'])

def add_item(link, desc, title, date):
    """
    Grab a page and extract what we need. Ignore it if it doesn't have any of
    our desired keywords.
    """
    item = rsslib.Item()
    item.link = link
    item.description = 'Keywords: ' + desc
    item.title = title
    item.pubDate = datetime.strptime(date, '%Y-%m-%d, %H:%M%p %Z')
    return item

# initialize the feed...
rss = rsslib.RSS()
rss.channel.link = JOBS_URL
rss.channel.title = "Craig's List jobs"
rss.channel.description = "Software jobs of interest at Craig's List"

# get the page from Craig's List...
html_doc = html.parse(JOBS_URL)

# loop through the anchors matching the city, extracting the hrefs. Parse those
# pages and get the fields we need...
item_list = []
for l in html_doc.xpath('//a[contains( @href, "%s")]/@href' % HREF_STR):
    _h = html.parse(l)
    body = _h.xpath('//body')[0].text_content()
    post_date = re.findall( 'Date:\s+(\S+,?\s+\S+\s+\S+)', body, re.IGNORECASE )[0]
    h2 = _h.xpath('//h2/text()')[0]
    keyword_hits = set([ hit for hit in body.split() if (hit.lower() in KEYWORDS) ])
    if (keyword_hits == set([])):
        continue
    rss.addItem(add_item(l, ', '.join(keyword_hits), h2, post_date))

# output the feed...
print "Content-type: application/rss+xml"
print
print rss.write()

File: craigslist_jobs/ruby/craigslist_jobs.rb
#!/usr/bin/env ruby -w

require 'rubygems'
require 'mechanize'
require 'nokogiri'
require 'uri'
require 'rss'
require 'set'

JOBS_URL    = 'http://phoenix.craigslist.org/sof/index.html'
HREF_SUBSTR = %r{phoenix\.craigslist\.org/evl/sof/}
JOB_KEYWORDS = %w[ perl ruby sql python postgres ]

mech = Mechanize.new

mech.get(JOBS_URL)

content = RSS::Maker.make('2.0') do |_feed|
  _feed.channel.title       = "Craig's List jobs"
  _feed.channel.link        = JOBS_URL
  _feed.channel.description = "Software jobs of interest at Craig's List"
  _feed.items.do_sort       = true # sort items by date

  mech.page.links_with(:href => HREF_SUBSTR).each do |_link|

    mech.get(_link.href)
    response_body = Nokogiri::HTML(mech.page.content).inner_text

    hits = response_body.split(' ').select{ |w| JOB_KEYWORDS.include?(w.downcase) }.to_set
    if (hits.any?)
      post_date = response_body[/Date:\s+(\S+\s+\S+\s+\S+)/i, 1]
      i = _feed.items.new_item
      i.date        = Time.parse(post_date)
      i.link        = _link.href
      i.title       = mech.page.at('h2').text.squeeze(' ').strip
      i.description = 'Matching words: ' << hits.to_a.join(', ')
    end
  end
end

puts "Content-type: application/rss+xml"
puts
puts content, "\n"
